[🥒Team-6 / iOS / 쏭 & 잭슨] 상세 페이지 및 Persistence 구현

1song2 opened the pull request:
상쾌한 월요일입니다 제이슨☀️
지난 PR 피드백 및 새로운 기능 구현을 코드에 반영해 2차 PR을 드립니다.
프로젝트 기간이 종료되고 PR 드리는만큼 제이슨 시간되실 때 가능하신만큼만 여유롭게 피드백 해주시면 좋을 것 같아요🙏
​
## 작업 목록
### 지난 PR 피드백 반영
- [x] `FetchDishesUseCase`의 RequestValue에서 불필요한 값을 제거 (Category 전체 대신 Category의 name 속성만 요청)
- [x] `DishCell`에서 네트워크 관련 logic 제거
- [x] `DishesListItemViewModel` 내부 속성을 `Dish` entity로 대체
- [ ] `reloadData` 대신 `reloadSections`로 특정 섹션의 데이터만 업데이트
- [ ] Test 코드 작성
### 기능 구현
- [x] 네트워크 요청으로 받아온 반찬 목록을 Persistence로 저장 (Realm 이용)
- [x] 반찬 배송 정보, 수량 선택, 상세 이미지 등을 대응할 수 있는 상세 페이지 구현
- [x] 메인 페이지에서 특정 셀을 탭하면 해당 상세 페이지로 이동
### 구조 설계 흐름도
- [x] 각 클래스를 Diagram화 & MVVM 계층에 맞게 분류 [링크](https://github.com/rla36/sidedish/wiki/iOS-앱-구조-및-흐름도)
​
## 학습 키워드
`Realm`, `UIScrollView (Paging)`, `reloadSections`
​
## 고민과 해결
### Realm 데이터 저장
from @JacksonPk
이번 프로젝트에서 Persistence를 이용하는 방법을 고안했고 DB를 사용하기위해 Realm을 써 보았습니다.
Alamofire의 Result값에 따라 items를 가져올 시 Realm에 dish값을 저장했고, 네트워크 연결이 안될 시(WIFI/LTE OFF) Realm에 있는 DB를 불러와 메인 페이지에 보여주는 방식으로 고안했습니다.
궁금했던 점은 Persistence에 있는 이 데이터들을 고안한 방식처럼 관리해도 괜찮은 것인지, 혹은 더 나은방법이나 다른 접근으로 사용할 수 있는지 궁금합니다.
​
### reloadSections
from @1song2
이번 프로젝트의 메인디쉬라고 말씀해주신 `reloadSections` 부분을 꼭 구현해보고 싶어서 지난 주말동안 시도해보았습니다.
결과적으로 성공하진 못했는데요😭 시도해보는 과정 중에서 느낀 점이 몇가지 있어 아래 공유드립니다.
#### 문제 상황 및 시도 방안
1. [문제] `items`는 section에 대한 정보를 가지고 있지 않기 때문에 `reloadSections(_:)`에 매개변수로 담을 section index를 알 수 없음
  * [시도] `category`에 `sectionIndex`와 `items`를 속성값으로 추가
  * `category`를 observe해 `items`에 변동 사항이 있을 시 해당 `sectionIndex`로 `reloadSections` 시도
2. [문제] 1번과 같이 수정 후 콘솔에 아래와 같은 에러 메시지 발생 (프로그램이 동작은 됨)
```
[UICollectionView] Invalid update: invalid number of items in section 0. 
The number of items contained in an existing section after the update (11)
must be equal to the number of items contained in that section before the update (0),
plus or minus the number of items inserted or deleted from that section (0 inserted, 0 deleted)
and plus or minus the number of items moved into or out of that section (0 moved in, 0 moved out). - will perform reloadData.
UICollectionView instance: <UICollectionView: 0x7fe58382de00; frame = (0 47; 390 763); clipsToBounds = YES;
autoresize = RM+BM; gestureRecognizers = <NSArray: 0x600003cf34e0>; layer = <CALayer: 0x6000032d77e0>; 
contentOffset: {0, 0}; contentSize: {390, 294}; adjustedContentInset: {0, 0, 0, 0}; layout: <UICollectionViewFlowLayout: 0x7fe582f10c00>;
dataSource: <BanchanCode.MainPageCollectionViewDataSource: 0x6000032898e0>>;
currentUpdate: [UICollectionViewUpdate - 0x7fe581c124d0: old:<UICollectionViewData: 0x600000d884b0> new<UICollectionViewData: 0x600000d8c000>
items:<(
    "DS(1)",
    "IS(1)"
)>]
```
  * [시도] 브레이크 포인트 등을 통해 파악한 결과, 두 section이 동시에 업데이트 되는 시점에 해당 에러가 발생함을 파악 (동시성 이슈?)
  * `reloadSections` 사용시, 현재 `reloadSections`를 시도하는 section 외에는 변동사항이 발생해서는 안됨을 이해
  * 해당 에러가 발생하면 Xcode 자체에서 `reloadSections` 대신 `reloadData`를 대신 호출
​
해당 문제를 해결한 다른팀 동료분들 코드를 참고해 수정하고자 시도는 해보았지만, 아무래도 어느 정도 개발이 진행된 코드의 전체 구조를 뜯어고쳐야 하는 부분이기에 쉽지가 않았습니다😢
* 수정을 시도했다 포기한 부분
  * main, soup, side를 각각의 view model로 두는 대신 하나의 view model로 묶고 내부에 배열 혹은 딕셔너리로 main, soup, side를 가지고 있도록 함
  * Observable 대신 closure를 사용해 items 값에 변화가 생기는 시점에 해당 카테고리 값을 view controller로 전달할 수 있도록 함


꼭 해보면 좋겠다고 팁을 주신 부분인데 완전히 구현하지 못하고 넘어가 너무 아쉽지만,
그래도 해결을 위해 여러가지로 시도해보면서 느끼고 배우는 게 참 많았습니다. 제이슨 감사드려요.
(말씀주시기 전까진 `reloadSections`라는 메소드가 존재하는지도 몰랐었습니다😭)
추후 기회가 된다면 반찬 프로젝트를 아예 처음부터 구현하면서 해당 기능을 다시 시도해보고 싶네요!
​
2주동안 정말 고생 많으셨고 감사했습니다. 새로운 한주, 한달 힘내서 시작하시길 바랄게요!

JacksonPk committed:
Update README.md

JacksonPk committed:
Update README.md

JacksonPk committed:
Update README.md

JacksonPk committed:
Merge branch 'dev-iOS' of https://github.com/rla36/sidedish into iOS/feat/Realm

JacksonPk committed:
feat : add Realm library(#24)

JacksonPk committed:
feat : add Realm library(#24)

1song2 committed:
feat : add navigation controller

1song2 committed:
feat : add navigation controller

1song2 committed:
feat : add navigation controller

1song2 committed:
feat : add navigation controller

1song2 committed:
feat : add DetailPageViewController

1song2 committed:
feat : add DetailPageViewController

1song2 committed:
feat : go to DetailPageViewController when specific cell is selected

1song2 committed:
refactor : rename method to goToDishDetail

1song2 committed:
refactor : rename method to goToDishDetail

1song2 committed:
refactor : replace mockup API URL with real one

1song2 committed:
feat : add thumbnailImageScrollView

1song2 committed:
feat : add thumbnailImageScrollView

JacksonPk committed:
feat : add dish data in DB

JacksonPk committed:
feat : add dish data in DB

1song2 committed:
feat : add required elements to DetailPageViewController

1song2 committed:
feat : add required elements to DetailPageViewController

1song2 committed:
feat : add DishDetail and DishDetailResponseDTO

1song2 committed:
feat : add DishDetail and DishDetailResponseDTO

1song2 committed:
feat : let ViewController get id information when cell is selected

1song2 committed:
feat : let ViewController get id information when cell is selected

1song2 committed:
feat : show detailed info on dish

1song2 committed:
feat : show detailed info on dish

1song2 committed:
feat : let ViewController get category information when cell is selected

1song2 committed:
feat : let ViewController get category information when cell is selected

1song2 committed:
feat : let ViewController get category information when cell is selected

JacksonPk committed:
fix : enable to add array to Realm by list<T>() (#24)

JacksonPk committed:
fix : enable to add array to Realm by list<T>() (#24)

1song2 committed:
feat : add FetchDishDetailsUseCase

1song2 committed:
feat : add FetchDishDetailsUseCase

1song2 committed:
refactor : replace category with categoryName

1song2 committed:
feat : show detailed information about selected dish

1song2 committed:
feat : show detailed information about selected dish

1song2 committed:
refactor : make labels multiline

1song2 committed:
refactor : make labels multiline

1song2 committed:
chore : update DetailPageView UI

1song2 committed:
chore : update DetailPageView UI

1song2 committed:
chore : update UI of point and delivery

1song2 committed:
chore : update UI of point and delivery

1song2 committed:
feat : add custom stepper

1song2 committed:
feat : add custom stepper

1song2 committed:
refactor : change value of Obervable to non-optional DishDetail type

1song2 committed:
refactor : change value of Obervable to non-optional DishDetail type

1song2 committed:
refactor : embed badgeBackgroundView and label in stack view

1song2 committed:
refactor : embed badgeBackgroundView and label in stack view

JacksonPk committed:
feat : add functions data to DB and get from DB (#24)

JacksonPk committed:
feat : add functions data to DB and get from DB (#24)

JacksonPk committed:
feat : add RealmManger to handle Realm

JacksonPk committed:
feat : add RealmManger to handle Realm

1song2 committed:
refactor : add thousand separator to prices

1song2 committed:
refactor : add thousand separator to prices

1song2 committed:
refactor : customize button style

1song2 committed:
refactor : customize button style

1song2 committed:
refactor : customize button style

1song2 committed:
refactor : add section inset

1song2 committed:
refactor : add section inset

1song2 committed:
refactor : add section inset

1song2 committed:
refactor : add section inset

1song2 committed:
refactor : add section inset

1song2 committed:
refactor : add section inset

1song2 committed:
refactor : add section inset

1song2 committed:
refactor : add section inset

JacksonPk committed:
refactor : rename get set methods

JacksonPk committed:
refactor : rename get set methods

1song2 committed:
refactor : change catogory type to non-observable

1song2 committed:
refactor : update stepper state

1song2 committed:
refactor : make thumbImage nil in prepareForReuse

1song2 committed:
refactor : make thumbImage nil in prepareForReuse

1song2 committed:
refactor : make thumbImage nil in prepareForReuse

1song2 committed:
refactor : make thumbImage nil in prepareForReuse

JacksonPk committed:
chore

JacksonPk committed:
chore

JacksonPk committed:
Merge pull request #25 from rla36/iOS/feat/detail-page-view

[iOS] DetailPageView 구현

JacksonPk committed:
Merge remote-tracking branch 'refs/remotes/banchancode/dev-iOS'

Conflicts:
	.DS_Store
	iOS/BanchanCode/BanchanCode/Domain/UseCases/FetchDishesUseCase.swift
	iOS/BanchanCode/BanchanCode/Presentation/MainPageViewController.swift
	iOS/BanchanCode/BanchanCode/Presentation/ViewModel/DishesItemViewModel.swift
	iOS/BanchanCode/BanchanCode/Presentation/ViewModel/DishesViewModel.swift
	iOS/BanchanCode/BanchanCode/Presentation/Views/DishCell.swift

JacksonPk committed:
feat : merge with iOS/feat/detail-page-view

JacksonPk committed:
feat : add DishDetailDB funcs

1song2 committed:
refactor : add custom UIView BadgeView

1song2 committed:
refactor : add custom UIView BadgeView

1song2 committed:
refactor : break DishDetail up into three elements

1song2 committed:
refactor : add KeyColors enum

1song2 committed:
refactor : add KeyColors enum

JacksonPk committed:
refactor : move add to Realm func from ViewModel to UseCase

JacksonPk committed:
refactor : add failure case to network functions to use RealmDB

JacksonPk committed:
refactor : move DBlayer in ViewModel to UseCase

JacksonPk committed:
chore : remove comments(#24)

1song2 committed:
refactor : move network related logic to ViewModel

1song2 committed:
refactor :extend NSAttributedString to make part of the text bold

1song2 committed:
refactor :extend NSAttributedString to make part of the text bold

1song2 committed:
refactor :extend NSAttributedString to make part of the text bold

1song2 committed:
refactor :extend NSAttributedString to make part of the text bold

1song2 committed:
style : add line breaks for better readability

1song2 committed:
style : add line breaks for better readability

1song2 committed:
refactor : modify method performRequest to generic function

1song2 committed:
style : remove unnecessary line break

1song2 committed:
style : remove unnecessary line break

1song2 committed:
style : remove unnecessary line break

1song2 committed:
style : remove unnecessary line break

1song2 committed:
refactor : move currentQuantity and totalPrice to ViewModel

1song2 committed:
refactor : move price related logic to ViewModel

1song2 committed:
refactor : change paramter type to optional Int

1song2 committed:
refactor : change paramter type to optional Int

1song2 committed:
refactor : change paramter type to optional Int

JacksonPk committed:
Merge remote-tracking branch 'refs/remotes/banchancode/iOS/feat/refactoring'

Conflicts:
	iOS/BanchanCode/BanchanCode/Domain/UseCases/FetchDishesUseCase.swift
	iOS/BanchanCode/BanchanCode/Network/NetworkManager.swift
	iOS/BanchanCode/BanchanCode/Presentation/DetailPageViewController.swift
	iOS/BanchanCode/BanchanCode/Presentation/ViewModel/DishDetailsViewModel.swift

JacksonPk committed:
refactor : merge with refactoring

JacksonPk committed:
refactor : update overall codes

1song2 committed:
Merge pull request #26 from rla36/iOS/feat/Realm

[iOS merge] 데모 전 임시 최종

JacksonPk committed:
feat : add sectionHeaderContent with hidden option(#17)

JacksonPk committed:
feat : add sectionHeaderContent with hidden option(#17)

JacksonPk committed:
Merge pull request #27 from rla36/iOS/feat/Toast

Merge to Ios/feat/toast

1song2 committed:
Merge pull request #28 from rla36/iOS/feat/toast

[iOS] 헤더 섹션 터치 시 상품수량 보여주기#17

1song2 committed:
refactor : try to reload each section

1song2 committed:
refactor : try to reload each section

1song2 committed:
refactor : move network related code from cell to data source

1song2 committed:
refactor : move network related code from cell to data source

1song2 committed:
refactor : move price related logic from cell to view model

1song2 committed:
refactor : move price related logic from cell to view model

1song2 committed:
Merge branch 'iOS/feat/refactoring' into dev-iOS

1song2 committed:
Merge branch 'iOS/feat/refactoring' into dev-iOS

1song2 committed:
Merge branch 'dev-iOS' of https://github.com/rla36/sidedish into dev-iOS

ehgud0670 commented:
넵 ㅎㅎ 고생하셨습니다. 천천히 살펴보고 코멘트 드릴게요!

1song2 committed:
chore : replace URL address with mockup URL

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Data/PersistentStorage/DishDB.swift:
```diff
@@ -0,0 +1,33 @@
+    let prices = List<Int>()
+    let badges = List<String>()
+
+    convenience init(id: Int, name: String, contents: String, imageURL: String, categoryName: String) {
```
`covenience init`을 쓰신 이유가 따로 있을까요? 특별한 이유가 없으시다면 그냥 init을 써도 될 것 같습니다.

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Data/PersistentStorage/DishDB.swift:
```diff
@@ -0,0 +1,33 @@
+    let badges = List<String>()
+
+    convenience init(id: Int, name: String, contents: String, imageURL: String, categoryName: String) {
+        self.init()
+        self.id = id
+        self.name = name
+        self.contents = contents
+        self.imageURL = imageURL
+        self.categoryName = categoryName
+    }
```
이 부분에서 prices 와 badges를 추가하는게 깔끔하겠네요. 그리고 찾아보니 `append(objectsIn:)` 라는 것도 있는데 써보시는 게 어떨까요?

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Presentation/MainPageViewController.swift:
```diff
@@ -6,44 +6,61 @@
-        }
+        }                
         
         mainPageDataSource?.viewModels = viewModels
+        mainPageDelegate?.viewModels = viewModels
         
```
mainVC에도 viewModels를 가지고 있고 datasource와 delegate에도 viewModels를 가지고 있네요. 
이렇게 여러 객체가 공통된 viewModels를 가지고 있는 것보단 **하나의 객체만 프로퍼티로 가지고 있는 것**이 더 깔끔하고 로직이 단순하여 코드를 파악하기 좋습니다.
최소한 이 프로젝트에서는 단 하나의 객체만이 가지게 할 수 있으니 여유 있으실때 시도해보세요.

1song2 replied:
지금 간단히 생각해보기로는 delegate나 dataSource를 별도의 객체로 분리하지 않고, `MainPageViewController`를 extend해 사용한다면 viewModels를 하나로 사용할 수 있지 않을까 싶은 생각이 드네요!

viewController를 가볍게 사용하고자 파일을 분리해 작성한 것이었는데, 말씀하신대로 여러 객체가 viewModels를 가지고 있는 것은 확실히 헷갈리고 깔끔하지 않은 것 같습니다🤔

또 다른 개선 방법이 있을지도 좀 더 고민해보겠습니다!

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Presentation/MainPageViewController.swift:
```diff
@@ -6,44 +6,61 @@
     @IBOutlet weak var dishCollectionView: UICollectionView!
     private var mainPageDelegate: MainPageCollectionViewDelegate?
     private var mainPageDataSource: MainPageCollectionViewDataSource?
+    private var viewModels: [DishesViewModel]!
```
viewModels 배열(자료구조)를 가지고 있는 CategoryViewModel 객체를 만드는 건 어떨까요? 
이러한 객체가 있으면 메소드로 더 추상화 가능할것 같아요. 

> 팁 
* 하나의 총체적인 뷰모델을 먼저 생각하고 안에 세부적인 뷰모델을 작성한다고 생각하시면 쉽게 쉽게 객체를 만드실 수 있을 것 같아요.
* 맨 처음에 뷰모델을 구현하실때 하나의 VC에서 제일 큰 뷰모델 하나를 만들고 프로퍼티로 모델 값만 두고 메소드만 추가하시는 걸 추천드립니다. 그 다음에 세부적인 뷰모델을 차근차근 구현하는거죠.

1song2 replied:
네 제이슨 주신 의견에 공감합니다!
자료를 많이 찾아본 것은 아니지만 이렇게 viewModel을 여러개 가지고 있기 보다는 크게 아우르는 하나의 viewModel을 사용하는 것이 일반적인 것 같다는 느낌을 받았어요.
다만 개선이 생각처럼 되지 않아 그냥 현재의 구조로 두고 아쉬움만 느끼고 있었는데,
팁 주신 내용을 참고해 한번 수정해보도록 하겠습니다!

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Presentation/MainPageViewController.swift:
```diff
@@ -60,11 +77,22 @@ class MainPageViewController: UIViewController {
     private func bind(to viewModel: DishesViewModel) {
-        viewModel.category.observe(on: self) { [weak self] _ in self?.updateItems() }
-        viewModel.items.observe(on: self) { [weak self] _ in self?.updateItems() }
+        viewModel.category.observe(on: self) { [weak self] category in
+            DispatchQueue.main.async {
+                self?.updateSection(at: category.sectionIndex)
+            }
+        }
```
observe 메소드가 category에서만 쓰이는 군요!
items의 observe 메소드가 쓰이지 않는다면 items는 `Observable<[DishesItemViewModel]>` 타입이 아닌 `[DishesItemViewModel]` 이 타입이어도 될 것 같네용

1song2 replied:
`reloadSections`를 시도해보는 과정에서 category를 급하게 추가하다 보니 필요 없는 부분에서 Observable이 사용되고 있는 걸 놓쳤네요. PR 전에 작성했던 코드를 좀더 꼼꼼히 살피는 습관을 들이도록 하겠습니다!

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Presentation/ViewModel/DishesViewModel.swift:
```diff
@@ -21,34 +26,41 @@ protocol DishesViewModel: DishesViewModelInput, DishesViewModelOutput { }
+        fetchDishesUseCase.execute(requestValue: .init(categoryName: category.value.name), completion: { result in
             switch result {
             case .success(let items):
+                self.category.value.items = items.dishes
                 self.items.value = items.dishes.map(DishesItemViewModel.init)
```
여기 보시면 items의 데이터 값은 갱신되지 않았는데 category의 모델 값이 먼저 갱신되어 reloadSections를 호출하고 있네요!

1song2 replied:
`DishesViewModel`이 category와 items를 따로 들고 있는데, category가 또 내부 속성으로 items를 들고 있는 지금의 구조 때문인 것 같습니다. 다시 보니 구조가 정말 요상하네요😵
위에서 피드백 주신 것처럼 `CategoryViewModel`을 만들면서 개선해보겠습니다!

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Presentation/Views/DishCell.swift:
```diff
@@ -15,43 +15,48 @@ class DishCell: UICollectionViewCell {
-        }
-        nameLabel.text = viewModel.name
-        descriptionLabel.text = viewModel.description
-        let prices = viewModel.prices
-        let originalPrice = prices[0]
-        if prices.count > 1 {
-            let lastPrice = prices[1]
-            originalPriceLabel.attributedText = "\(originalPrice)원".strikethrough()
-            lastPriceLabel.text = "\(lastPrice)원"
+        let dish = viewModel.dish
+        nameLabel.text = dish.name
+        descriptionLabel.text = dish.description
+        
+        lastPriceLabel.text = String().format(price: viewModel.lastPrice)
+        if let originalPrice = viewModel.originalPrice {
+            originalPriceLabel.attributedText = String().format(price: originalPrice)?.strikethrough()
+            originalPriceLabel.isHidden = false
         } else {
-            lastPriceLabel.text = "\(originalPrice)원"
```
뷰모델의 존재 이유는 사용하는 메소드들을 테스트하기 위함도 있습니다. 
```swift
String().format(price: originalPrice)?.strikethrough()
```
=> 이러한 코드를 뷰모델안에 메소드로 추상화하고 이 메소드를 테스트하면 정말 아름답겠죠? ☺️

1song2 replied:
JK도 테스트를 하지 않는다면 뷰모델을 사용하는 메리트가 없다고 종종 말씀해주시더라구요🤭
아직 버릇이 들지 않아 테스트를 등한시 하는 경향이 있는데 앞으로는 좀더 이부분을 보강해보도록 하겠습니다🔥

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Presentation/Views/DishCell.swift:
```diff
@@ -15,43 +15,48 @@ class DishCell: UICollectionViewCell {
         }
     }
+    
+    override func prepareForReuse() {
```
해당 함수 잘 사용하셨습니다. 👍

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Network/NetworkManager.swift:
```diff
@@ -9,29 +9,29 @@ import Foundation
         AF.request(urlString, method: .get)
             .validate(statusCode: 200..<300)
-            .responseDecodable(of: DishesResponseDTO.self) { (response) in
+            .responseDecodable(of: T.self) { response in
```
responseDecodable의 클로져 내용을 메인 스레드에서 처리하네요!
UI 렌더링하는 코드가 아니라면 다른 스레드에서 클로저를 처리하도록 함수 매개변수로 queue를 추가하는 건 어떨까요?
네트워크 처리, response 처리 이러한 것 모두 메인 스레드가 아닌 다른 스레드에서 하도록 하면 좋습니다.

1song2 replied:
와.. 피드백 주신 내용 처음엔 이해를 못하고
'`performRequest`를 사용하는 caller에 `DispatchQueue.main.async { }`를 쓰는 부분이 있다는 말씀인가?'하고 물음표만 띄웠었는데요,
`responseDecodable`이 디폴트로 main에서 스레드를 처리한다는 말씀이셨군요.. 대박😭
```
The queue on which the completion handler is dispatched. .main by default.
```

별 생각 없이 디코딩까지 한번에 해준다니 넘 편하다고 갖다쓴 메소드인데 설명을 꼼꼼히 읽지 못한 점을 반성하게 되네요.
처음 사용하거나 잘 모르는 부분은 코드에 도입하기 전, 좀더 면밀한 검토가 없으면 생각했던 것과 다른 결과가 나올 수 있다는걸 배웠습니다🙇🏻‍♀️

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Presentation/MainPageViewController.swift:
```diff
@@ -60,11 +77,22 @@ class MainPageViewController: UIViewController {
-        viewModel.category.observe(on: self) { [weak self] _ in self?.updateItems() }
-        viewModel.items.observe(on: self) { [weak self] _ in self?.updateItems() }
+        viewModel.category.observe(on: self) { [weak self] category in
+            DispatchQueue.main.async {
```
`DispatchQueue.main.async` 이 부분 주석처리해도 UI를 메인스레드에서 처리하고 있지 않다는 경고가 안뜰껍니다. 왜 그런지 고민해보세요 ㅎㅎ

1song2 replied:
위에 피드백 주셨던 `responseDecodable`부터 계속 메인에서 처리를 하고 있기 때문일까요?👀

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Presentation/MainPageViewController.swift:
```diff
@@ -60,11 +77,22 @@ class MainPageViewController: UIViewController {
+        }
+    }
+    
+    private func updateSection(at index: Int) {
+        self.dishCollectionView.reloadSections(IndexSet(integer: index))
     }
```
DispatchQueue.main.async를 여기에 옮겨도 되지 않을까요?

1song2 replied:
아래 DispatchQueue 관련 주셨던 다른 피드백과 종합해서 생각해보면,
좀더 직접적으로 UI를 처리해주는 부분에서 DispatchQueue를 사용하는 것이 낫다는 말씀이실까요?
아직 완벽히 감이 오진 않아서 아래 말씀해주신 것처럼 스레드간 역할을 좀더 학습해보고 그부분 고민해 코드 작성해보겠습니다!

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Data/PersistentStorage/RealmManager.swift:
```diff
@@ -0,0 +1,47 @@
+}
+
+class RealmManager: RealmOperations {
+    let realm = try! Realm()
```
forcedunwrapping 은 좋지 않아요. 아예 사용 안하시는 걸 권장합니다.
try? 을 사용하거나 do-try-catch 구문으로 에러를 잡는 것이 좋습니다

JacksonPk replied:
감사합니다. 튜토리얼대로 따라만 하다보니 force를 이용했습니다. 앞으로는 try catch로 구현해보는 습관을 키우도록 하겠습니다!

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Data/PersistentStorage/RealmManager.swift:
```diff
@@ -0,0 +1,47 @@
+                item.dish.badges.forEach {
+                    dishDB.badges.append($0)
+                }
+                try! realm.write {
+                    realm.add(dishDB)
+                }
```
* 마찬가지로 try? 을 사용하거나 do-try-catch 구문으로 에러를 잡는 것이 좋습니다. 
* 또 스위프트에서 Error란 처리가능한 에러를 말합니다. 스위프트의 Error는 자바, 오브젝티브씨의 Error 및 Exception와는 또 어떻게 다른지 공부하시면 Swift를 보다 깊게 이해하실 수 있을 것 같아요.
  *  이 부분에 대해서는 제가 스터디하면서 작성한 [글](https://github.com/TheSwiftists/effective-swift/blob/item71/10%EC%9E%A5_%EC%98%88%EC%99%B8/item71.md)이 있는데 한번 보시는 것도 좋을것 같아요 ㅋㅋㅋ ㅎㅎ. 비판은 환영입니다!!

JacksonPk replied:
감사합니다. 이번 기회에 저도 Result<T,Error> 에 대한 개념을 살짝 맛보았는데( 네트워크 연결이 안됐을 때 에러 처리를 이용해서 DB의 아이템을 불러오는 방식으로 구현하기 위해서) 제이슨이 써주신 글을 한번 읽어보고 이해해보겠습니다!

ehgud0670 replied:
아 넵, 글이 좀 정제되어 있진 않아요 ㅋㅋㅋ

ehgud0670 commented on iOS/BanchanCode/BanchanCode/Presentation/DetailPageViewController.swift:
```diff
@@ -0,0 +1,162 @@
+        }
+    }
+    
+    private func refreshView() {
```
DispatchQueue.main.async가 observe에 있는것보다 refreshView처럼 **UI처리하는 메소드에 위치**하는게 자리잡는 게 더 낫지 않을까요? 메인 스레드가 다른 스레드와 다르게 어떤 역할을 할 수 있는지 고민하시고 코드 작성하시면 좋을 것 같습니당

ehgud0670 reviewed (commented):
> 궁금했던 점은 Persistence에 있는 이 데이터들을 고안한 방식처럼 관리해도 괜찮은 것인지, 혹은 더 나은방법이나 다른 접근으로 사용할 수 있는지 궁금합니다.

​=>넵, 지금과 같이 작성해도 괜찮을 것 같네요! primaryKey도 잘쓰고 계시고요. 이 프로젝트 말고 더 복잡한 상황이 있어야 더 나은 방법이 뭐가 있을지 고민할 수 있을 것 같습니다. 다른 백엔드분들에게 DB 작성하기 까다로운 상황으로 어떤게 있는지 질문해보세요 ㅎㅎ
=> 다만 **사용하는 전체적인 로직이 효율적이지 못한 것 같습니다.** 모바일에서 DB는 네트워크 사용량을 줄이기 위해서도 사용합니다. 맨 처음에는 데이터를 서버로부터 받아 DB에 저장하고, 그 이후에는 DB에 데이터가 있다면 데이터를 서버로 요청하지 않고 DB의 데이터를 사용하는 방법은 어떨까요? 그리고  업데이트 Interval를 둬서 시간이 지나 데이터를 업데이트 해야 하면 다시 서버로부터 데이터를 요청하는거죠.

> 이번 프로젝트의 메인디쉬라고 말씀해주신 reloadSections 부분을 꼭 구현해보고 싶어서 지난 주말동안 시도해보았습니다.

=> 이 부분 구현해 보시려고 고생 많으셨습니다. 구현하려고 주말에도 코딩하시고 정말 대단하시네요 💯 저도 이 부분 구현 못해서 옆에 친구한테도 물어보고 그랬어요 ㅠㅠ ㅋㅋ. 그리고 이전 기수들 중에서도 이거 다시 구현해보라고 하면 못할 사람들도 꽤 있을거에요. 어려운 부분이니 너무 상심하지 않으셨으면 좋겠습니다. 그리고 reloadSections는 이전에 안썼으면 당연히 모르는 거니깐 신경쓰지마세요 ㅎㅎ
=> reloadSections 이 부분 그래서 지금 코드로 제가 구현해보았습니다. 이 부분 따로 오프라인으로 제가 학원에 가서 설명 해드릴수 있을 것 같은데 설명해드릴까요? 글로만 설명해드리려고 하니 복잡하네요 ㅋ.ㅋ 뵐 수 있으면 다른 부분들도 추가로 더 설명드리겠습니다! 

* 추가로 짧은 시간안에 DetailPageVC까지 구현하시고 대단하시네요! 고생많으셨습니다. 👏

p.s 너무 늦게 피드백을 드려 죄송하네요... ㅜ.ㅜ 반영은 안하셔도 되니 피드백만 봐주시고 코멘트 달아주세요 ㅎㅎ. 두 분도 피드백 확인하셨으면 머지 하셔도 됩니다.

JacksonPk commented:
> > 궁금했던 점은 Persistence에 있는 이 데이터들을 고안한 방식처럼 관리해도 괜찮은 것인지, 혹은 더 나은방법이나 다른 접근으로 사용할 수 있는지 궁금합니다.
> 
> ​=>넵, 지금과 같이 작성해도 괜찮을 것 같네요! primaryKey도 잘쓰고 계시고요. 이 프로젝트 말고 더 복잡한 상황이 있어야 더 나은 방법이 뭐가 있을지 고민할 수 있을 것 같습니다. 다른 백엔드분들에게 DB 작성하기 까다로운 상황으로 어떤게 있는지 질문해보세요 ㅎㅎ
> => 다만 **사용하는 전체적인 로직이 효율적이지 못한 것 같습니다.** 모바일에서 DB는 네트워크 사용량을 줄이기 위해서도 사용합니다. 맨 처음에는 데이터를 서버로부터 받아 DB에 저장하고, 그 이후에는 DB에 데이터가 있다면 데이터를 서버로 요청하지 않고 DB의 데이터를 사용하는 방법은 어떨까요? 그리고 업데이트 Interval를 둬서 시간이 지나 데이터를 업데이트 해야 하면 다시 서버로부터 데이터를 요청하는거죠.
> 

네, 그 로직이 좋은 방식이라고 생각합니다. 제가 계속 생각해봤던 것이 실제 앱이라고 생각해봤을 때, 서버에서 업데이트 되지 않았을까(예를들어, 특정 가게에서 메뉴의 가격을 바꾼다던지.) DB에 있는 데이터를 보여주면 사용자측면에서 혼돈이 올까 싶어서 네트워크가 연결이 되지 않을 경우에만 일시적으로 DB에 있는 데이터를 보여주도록 로직을 짜 보았습니다. 업데이트 Interval에 대한 개념을 이해하지못해서 이렇게 로직을 구현했네요. 감사합니다.

ehgud0670 commented:
넵, 추가로 time interval에 대해서도 말씀드리고자 코멘트 드렸어요, 수고하셨습니다! 👏

1song2 committed:
fix: change order of lines

items가 먼저 업데이트된 다음 category에 속한 items가 업데이트 되도록 순서를 조정

1song2 committed:
refactor: add queue as argument to method

1song2 committed:
fix: remove unnecessary code from methods
